pkg/registry/core/service/storage/rest_test.go

```diff
@@ -5086,6 +5087,109 @@ func TestDefaultingValidation(t *testing.T) {
 			expectedIPFamilies:     nil,
 			expectError:            true,
 		},
+
+		// preferDualStack services should not be updated
+		// to match cluster config if the user didn't change any
+		// ClusterIPs related fields
+		{
+			name:       "unchanged preferDualStack-1-ClusterUpgraded",
+			modifyRest: fnMakeDualStackStackIPv4IPv6Allocator,
+			oldSvc: &api.Service{
+				Spec: api.ServiceSpec{
+					Type:           api.ServiceTypeClusterIP,
+					ClusterIPs:     []string{"1.1.1.1"},
+					IPFamilies:     []api.IPFamily{api.IPv4Protocol},
+					IPFamilyPolicy: &preferDualStack,
+				},
+			},
+
+			svc: api.Service{
+				Spec: api.ServiceSpec{
+					Type:           api.ServiceTypeClusterIP,
+					ClusterIPs:     []string{"1.1.1.1"},
+					IPFamilies:     []api.IPFamily{api.IPv4Protocol},
+					IPFamilyPolicy: &preferDualStack,
+				},
+			},
+			expectedIPFamilyPolicy: &preferDualStack,
+			expectedIPFamilies:     []api.IPFamily{api.IPv4Protocol},
+			expectError:            false,
+		},
+
+		{
+			name:       "unchanged preferDualStack-2-ClusterDowngraded",
+			modifyRest: fnMakeSingleStackIPv4Allocator,
+			oldSvc: &api.Service{
+				Spec: api.ServiceSpec{
+					Type:           api.ServiceTypeClusterIP,
+					ClusterIPs:     []string{"1.1.1.1", "2001::1"},
+					IPFamilies:     []api.IPFamily{api.IPv4Protocol, api.IPv6Protocol},
+					IPFamilyPolicy: &preferDualStack,
+				},
+			},
+
+			svc: api.Service{
+				Spec: api.ServiceSpec{
+					Type:           api.ServiceTypeClusterIP,
+					ClusterIPs:     []string{"1.1.1.1", "2001::1"},
+					IPFamilies:     []api.IPFamily{api.IPv4Protocol, api.IPv6Protocol},
+					IPFamilyPolicy: &preferDualStack,
+				},
+			},
+			expectedIPFamilyPolicy: &preferDualStack,
+			expectedIPFamilies:     []api.IPFamily{api.IPv4Protocol, api.IPv6Protocol},
```

Comment:
The downgrade is the more complex thing, we end with a dual-stack service in a single stack cluster.
The funny thing is that I think that the IPv6 family may still work ... :)

Reply:
This is no different from changing service CIDR. services with allocated IP from wrong family will always be there until user fixes them.

Reply:
are they not fixed by the repair loop?

for downgrade the problem seems that the repair loop will only run with the configured cluster cidr, leaving the second one as is ... but I'm talking from the top of my head, I didn't have time to check it

Reply:
If we find IPs allocated outside the known CIDRs, what are we supposed to do?  We can't just change them (I think?) and we can't just delete them.  I think the root of the discussion is that changing CIDRs, especially deactivating them, needs to be carefully managed.

In this regard, your IP ranges change is superior - we can at least put a finalizer on it :)

Reply:
no fix. it will loop will generate an event for the user to fix. Which usually means delete/re-create due to immutability rules. Hence changing service-cidr is generally a bad idea.

Reply:
Even if we allowed `clusterIPs[0]` to change, we could still end up with Services that should not be possible.  E.g.

* create a PreferDualStack, get v4
* upgrade to dual-stack cluster
* change to RequireDualStack, get v4,v6
* downgrade to single-stack

That Service is RequireDual.  We can't automatically change that to Prefer.   If you created a new service with that config it would be rejected.  With the finalizer, the CIDR it draws from would be stuck in delete-pending, so at least we know something is wrong.

Reply:
Let us assume it is a single stack with ClusterIP outside the range (as in service-cidr has changed cluster wide). Today we let the repair loop deal with it (broadcasts an event) and user must delete and recreate it. Same applies for dual-stack. 

am i misunderstanding something?

Reply:
if you move from dual to single, you will only run the repair in the single clusterIP cidr family, the existent secondary ClusterIPs  would not go through the repair loop, right?

Reply:
no, repair loop has been modified to work with IPs in `spec.ClusterIPs` as part of the dual stack work.

Reply:
awesome , I see it now
https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/pkg/registry/core/service/ipallocator/controller/repair.go#L221-L226